File: src/cms-api/src/main/java/com/soen490/cms/Models/Course.java
// MIT License

// Copyright (c) 2019 teamCMS

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package com.soen490.cms.Models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import lombok.Data;
import javax.persistence.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@Entity
@Data
public class Course {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    private String name;

    private int number;

    private String title;

    private double credits;

    private String note;

    private int level;

    private double lectureHours;

    private double tutorialHours;

    private double labHours;

    @Lob
    private String description;

    @Lob
    private byte[] outline;

    private int isActive;

    @JsonIgnoreProperties({"courses", "degrees"})
    @ManyToOne
    @JoinColumn(name = "program_id")
    private Program program;

    @JsonManagedReference
    @OneToMany(mappedBy = "course")
    private List<Requisite> requisites = new ArrayList<>();

    @JsonIgnoreProperties("course")
    @OneToMany(mappedBy = "course")
    private List<DegreeRequirement> degreeRequirements = new ArrayList<>();

}

File: src/cms-client/src/app/models/course.ts
// MIT License

// Copyright (c) 2019 teamCMS

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import { Requisite } from '../models/requisite';
import { Program } from './program';

export class Course {
  id: number;
  credits: number;
  degreeRequirements: Object[];
  description: string;
  equivalent: string[];
  isActive: boolean;
  labHours: number;
  lectureHours: number;
  level: number;
  name: string;
  note: string;
  number: number;
  program: Object;
  requisites: Requisite[];
  title: string;
  tutorialHours: number;

  constructor(){
    this.id = 0;
    this.credits = 0;
    this.degreeRequirements = [];
    this.description = '';
    this.equivalent = [];
    this.isActive = false;
    this.labHours = 0;
    this.lectureHours = 0;
    this.level = 1;
    this.name = '';
    this.note = '';
    this.number = 100;
    this.program = new Program();
    this.requisites = [];
    this.title = '';
    this.tutorialHours = 0;
  }
}
